L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+      // store lost tracks
+      put(lost, lostStubs, lostTracks, channel);
+    }
+  }
+
+  // adds a layer to states
+  void KalmanFilter::layer(deque<State*>& stream) {
+    static constexpr int latency = 5;
+    // dynamic state container for clock accurate emulation
+    deque<State*> streamOutput;
+    deque<State*> stack;
+    // static delay container
+    vector<State*> delay(latency, nullptr);
+    delay.reserve(latency + 1);
+    // each trip corresponds to a f/w clock tick
+    // done if no states to process left, taking as much time as needed
+    while (!stream.empty() || !stack.empty() || !all_of(delay.begin(), delay.end(), [](const State* state){ return state == nullptr; })) {
+      State* state = pop_front(stream);
+      if (!state)
+        state = pop_front(stack);
+      streamOutput.push_back(state);
+      if (!state || !state->stub() || state->layer() != layer_)
+        state = nullptr;
+      if (state != nullptr)
+        comb(state);
+      delay.push_back(state);
+      state = pop_front(delay);
+      if (state != nullptr)
+        stack.push_back(state);
+    }
+    stream = streamOutput;
+    for (State*& state : stream) {
+      if (!state || !state->stub() || state->layer() != layer_ )
+        continue;
+      update(state);
+    }
+  }
+
+  // repicks combinatoric stubs for state
+  void KalmanFilter::comb(State*& state) {
+    // picks next stub on layer
+    StubKFin* stub = state->stub();
+    const int layer = stub->layer();
+    TrackKFin* track = state->track();
+    const vector<StubKFin*>& stubs = track->layerStubs(layer);
+    const TTBV& hitPattern = state->hitPattern();
+    const int pos = distance(stubs.begin(), find(stubs.begin(), stubs.end(), stub)) + 1;
+    StubKFin* stubNext = nullptr;
+    if (pos != (int)stubs.size())
+      stubNext = stubs[pos];
+    // picks next stub on different layer, nullifies state if skipping layer is not valid
+    else {
+      bool valid(true);
+      // having already maximum number of added layers
+      if (hitPattern.count() == setup_->kfMaxLayers())
+        valid = false;
+      // not enough layers remain after skipping
+      if (hitPattern.count() + track->hitPattern().count(layer + 1, setup_->numLayers()) < setup_->kfMaxLayers())
+        valid = false;
+      if (valid) {
+        // pick next stub on next populated layer
+        for (int nextLayer = layer_ + 1; nextLayer < setup_->numLayers(); nextLayer++) {
+          if (track->hitPattern(nextLayer)) {
+            stubNext = track->layerStub(nextLayer);
+            break;
+          }
+        }
+      }
+    }
+    if (stubNext) {
+      states_.emplace_back(state, stubNext);
+      state = &states_.back();
+    } else
+      state = nullptr;
+  }
+
+  // best state selection
+  void KalmanFilter::accumulator(deque<State*>& stream) {
+    // accumulator delivers contigious stream of best state per track
+    // remove gaps and not final states
```

Comment:
Can states which are "not final" actually reach the accumulator? And can a state with less than required number of layers reach it?

Reply:
not sure